backend/routes.js

```diff
@@ -63,6 +64,61 @@ const cidLookupTable = {
   [multicodes.SHELTER_FILE_CHUNK]: 'application/vnd.shelter.filechunk+octet-stream'
 }
 
+const limiterKey = (ip: string) => {
+  const ipVersion = isIP(ip)
+  if (ipVersion === 4) {
+    return ip
+  } else if (ipVersion === 6) {
+    // Likely IPv6
+    const [address, zoneIdx] = ip.split('%')
+    const segments = address.split(':')
+
+    // Is this a compressed form IPv6 address?
+    let isCompressed = false
+    for (let i = 0; i < segments.length - 1; i++) {
+      // Compressed form address
+      if (!isCompressed && segments[i] === '') {
+        const requiredSegments = 8 - (segments.length - 1)
+        if (requiredSegments < 0) {
+          throw new Error('Invalid IPv6 address: too many segments')
+        }
+        if ((i === 0 || i === segments.length - 2) && segments[i + 1] === '') {
+          segments[i + 1] = '0'
+        }
+        if (i === 0 && segments.length === 3 && segments[i + 2] === '') {
+          segments[i + 2] = '0'
+        }
+        segments.splice(i, 1, ...new Array(requiredSegments).fill('0'))
+        isCompressed = true
+        continue
+      }
+      // Remove leading zeroes
+      segments[i] = segments[i].replace(/^0+/, '0')
+    }
+
+    if (segments.length === 8 && isIP(segments[7]) === 4) {
+      // IPv4-embedded, IPv4-mapped and IPv4-translated addresses are returned
+      // as IPv4
+      return segments[7]
+    } else if (segments.length === 8) {
+      if (zoneIdx) {
+        segments[7] = segments[7].replace(/^0+/, '0')
+        // Use tagged (link-local) addresses in full
+        return segments.join(':').toLowerCase() + '%' + zoneIdx
+      } else {
+        // If an IPv6 address, return the first 64 bits. This is because that's
+        // the smallest possible subnet, and spammers can easily get an entire
+        // /64
+        return segments.slice(0, 4).join(':').toLowerCase() + '::'
+      }
+    } else {
+      throw new Error('Invalid IPv6 address')
+    }
+  }
+
+  throw new Error('Invalid address format')
+}
```

Comment:
OK, I do not understand why this code exists at all. I read the comment linked to in `// See discussion: https://github.com/okTurtles/group-income/pull/2280#pullrequestreview-2219347378` but it didn't really address this AFAICT.

Why are you doing all of this processing for ipv6 IPs? Why are you creating a `limiterKey` function? Why does this PR exist at all? I notice that no issue is referenced and therefore no problem is referenced, and I genuinely don't understand why the code should be modified.

Could you please update the code with a comment that addresses these questions?

Reply:
Sure, I can create an issue, but I'm not sure what exactly isn't being addressed in the linked comment. (See also https://github.com/okTurtles/group-income/pull/2280#discussion_r1704357074 and your comment acknowledging this https://github.com/okTurtles/group-income/pull/2280#discussion_r1704389923)

The existing code will use whatever's given as IP as a limiter key. The linked comment says:

> d. At the very least, the key used should be the /64 block for IPv6 addresses. Not doing this makes this PR pointless and exposes the server to very easy DoS.

This PR implements using the /64 block for IPv6 addresses.